code/parse/sexp.cpp

```diff
@@ -23298,6 +23567,706 @@ void sexp_set_motion_debris(int node)
 	Motion_debris_override = is_sexp_true(node);
 }
 
+/**
+* Check if a SEXP list container contains specific elements
+*/
+int sexp_list_has_data(int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name);
+		return SEXP_FALSE;
+	}
+
+	if (!Sexp_containers[index].is_list()) {
+		Warning(LOCATION, "List-has-data called on non-list container %s.", container_name);
+		return SEXP_FALSE;
+	}
+
+	const auto &list_data = Sexp_containers[index].list_data;
+
+	node = CDR(node);
+	// there should be at least one string to search for
+	Assert(node != -1);
+
+	SCP_string possible_data;
+	while (node != -1) {
+		// TODO: revisit if traversing the list for every item proves unacceptably slow
+		possible_data = CTEXT(node);
+
+		if (std::find(list_data.begin(), list_data.end(), possible_data) == list_data.end()) {
+			return SEXP_FALSE;
+		}
+
+		node = CDR(node);
+	}
+
+	return SEXP_TRUE;
+}
+
+/**
+*	SEXP which returns the index in a SEXP list container for a specific element.
+*	Returns -1 if the element is not found.
+*/
+int sexp_list_data_index(int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name);
+		return -1;
+	}
+
+	if (!Sexp_containers[index].is_list()) {
+		Warning(LOCATION, "List-data-index called on non-list container %s.", container_name);
+		return -1;
+	}
+
+	node = CDR(node);
+	Assert(node != -1);
+
+	const SCP_string possible_data = CTEXT(node);
+
+	int data_index = 0;
+	for (const SCP_string &data : Sexp_containers[index].list_data) {
+		if (possible_data == data) {
+			return data_index;
+		}
+		++data_index;
+	}
+
+	return -1;
```

Comment:
Should we issue a Warning() if we reach this line? Doesn't it indicate that the FREDder messed up? Or not necessarily?

Reply:
My reasoning is that if the FREDder just wants to know whether the list has some data, they would use the `list-has-data` SEXP. With `list-data-index`, the assumption is that the data is in the list somewhere and the FREDder just wants to know where in the list it is.

Reply:
I figured that they might wish to try either method. If they don't want the index but just want to know the data is still in there somewhere they'd use the list-has-data. If they actually want the index they'd use list-data-index and then test if the index is -1. That way they'd save having to search the container twice.